experimental/internal/taxa/names.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package taxa
+
+var (
+	// names is an array of user-visible names of all of the productions in this
+	// package.
+	names = [...]string{
+		Unknown:      "<unknown>",
+		Unrecognized: "unrecognized token",
+		TopLevel:     "file scope",
+		EOF:          "end-of-file",
+
+		Decl:         "declaration",
+		Empty:        "empty declaration",
+		Syntax:       "`syntax` declaration",
+		Edition:      "`edition` declaration",
+		Package:      "`package` declaration",
+		Import:       "import",
+		WeakImport:   "weak import",
+		PublicImport: "public import",
+		Extensions:   "extension range",
+		Reserved:     "reserved range",
+		Body:         "definition body",
+
+		Def:     "definition",
+		Message: "message definition",
+		Enum:    "enum definition",
+		Service: "service definition",
+		Extend:  "message extension block",
+		Oneof:   "oneof definition",
+
+		Option:       "option setting",
+		CustomOption: "custom option setting",
+
+		Field:     "message field",
+		EnumValue: "enum value",
+		Method:    "service method",
+
+		CompactOptions: "compact options",
+		MethodIns:      "method parameter list",
+		MethodOuts:     "method return type",
+
+		FieldTag:    "message field tag",
+		OptionValue: "option setting value",
+
+		Path:            "path",
```

Comment:
Aside: If we're going to use this in user-facing error messages, we really should update the language spec to make use of the term or at least define it.

It's slightly worrisome that this is so vague -- I worry the resulting error messages will not be as clear as saying "identifier", "type name", "option name", etc. I guess this is the one down-side to using such a lenient parser and AST model is that ascertaining the user's intent for a node in the AST is no longer trivial...

Reply:
I'm gonna rename this to "name", and add a separate noun for "type name", so that we can be more precise here.